flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/functions/ScalarFunctions.scala

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.table.runtime.functions
+
+import scala.annotation.varargs
+import java.math.{BigDecimal => JBigDecimal}
+import java.lang.{StringBuffer => JStringBuffer}
+
+/**
+  * All build-in scalar scalar functions.
+  */
+class ScalarFunctions {}
+
+object ScalarFunctions {
+
+  def power(a: Double, b: JBigDecimal): Double = {
+    Math.pow(a, b.doubleValue())
+  }
+
+  /**
+    * Returns the string that results from concatenating the arguments.
+    * Returns NULL if any argument is NULL.
+    */
+  @varargs
+  def concat(args: String*): String = {
+    val sb = new JStringBuffer
+    var i = 0
+    while (i < args.length) {
+      if (args(i) == null) {
+        return null
+      }
+      sb.append(args(i))
+      i += 1
+    }
+    sb.toString
+  }
+
+  /**
+    * Returns the string that results from concatenating the arguments and separator.
+    * Returns NULL If the separator is NULL.
+    *
+    * Note: CONCAT_WS() does not skip empty strings. However, it does skip any NULL values after
+    * the separator argument.
+    *
+    * @param args The first element of argument is the separator for the rest of the arguments.
+    */
+  @varargs
+  def concat_ws(args: String*): String = {
+    val separator = args(0)
+    if (null == separator) {
+      return null
+    }
+
+    val sb = new JStringBuffer
+
+    var i = 1
+    val dataList = args.filter(null != _)
```

Comment:
Why not doing the filtering within the while loop? You are iterating over the data twice. We should not use fancy Scala functions within runtime code.